### ISSUE TITLE
visited links auditing/plugin&mimetypes

### ISSUE THREAD

**Issue by L-a-n-g-o-l-i-e-r-s** [gorhill/uMatrix#488]
Hi there @gorhill, I was wondering if the current implementation of uMatrix or uBlock Origin provides a preference for layout.css.visited_links_enabled false/true?

This extension provides some information on the subject:
https://addons.mozilla.org/en-US/firefox/addon/disable-visited-links-css/ I am unsure if this is actually being used in the wild or not but it never hurts.

A few other questions for you, I know that uMatrix has control of some kind over plugin announcements, does this also include mimetypes? Could you explain how it goes about blocking this given that Mozilla has removed the preference I thought responsible for this behavior? See: https://groups.google.com/forum/#!topic/mozilla.dev.tech.plugins/_PsjN5SvZlo I was using this extension to configure this functionality: https://github.com/ChrisAntaki/plugins-and-mimetypes-firefox it was subsequently taken down because of the feature removal. (News to me.) 

https://github.com/gorhill/uMatrix/issues/282

**Comment by gorhill** [comment]
> provides a preference for layout.css.visited_links_enabled false/true?

No. I do not plan supporting this, to far removed from dealing with network requests.

> control of some kind over plugin announcements

I don't understand that part.

**Comment by berrythesoftwarecodeprogrammar** [comment]
just put those in your user.js. maybe check out https://github.com/pyllyukko/user.js

firefox removed the preference which dealt with plugin enumeration but if you have noscript installed you can make the following preferences:
`noscript.surrogate.noplugin.exceptions` = a whitelist of websites to allow plugin enumeration on (when they detect flash that way or something e.g. `.abc.net.au .domain.com`
`noscript.surrogate.noplugin.replacement` = `Object.defineProperty(navigator, \"plugins\", {value: []});`
`noscript.surrogate.noplugin.sources` = `@^https?://`

**Comment by L-a-n-g-o-l-i-e-r-s** [comment]
@gorhill I mean, when plugins are not allowed for a website via uMatrix and you have ask to activate enabled - you won't be prompted to activate that plugin.

If you allow plugins for the domain with uMatrix, you are then asked if you want to activate that plugin on x site for y content.  How is this being accomplished when what I thought was responsible for this functionality was removed?

Does that make things clearer, I can't really think of a better way to put it. 

@berrythesoftwarecodeprogrammar thanks for your reply, I will consider this, is user.js a firefox profile object? I am not interested in being in privacy mode 24/7 however, I didn't read too far but I hope there's an option to not do that. (Yes, I am now far from the scope of my question, apologies.)

Interesting idea about the userscripts, you have tested that as valid and working?

**Comment by berrythesoftwarecodeprogrammar** [comment]
go to your firefox profile and you will see prefs.js, and possibly user.js. prefs.js is where all about:config entries are stored and it shouldnt be edited manually. you can just add/edit those preferences in about:config OR make a file called user.js and add the following lines:
```
user_pref("noscript.surrogate.noplugin.exceptions",".abc.net.au .dailyliked.net");
user_pref("noscript.surrogate.noplugin.replacement","Object.defineProperty(navigator, \"plugins\", {value: []});");
user_pref("noscript.surrogate.noplugin.sources","@^https?://");
```
its not a userscript. its a file which firefox reads when it starts up and loads the entries into about:config. its a way to make sure you have some static preferences which dont get changed easily as the browser will set them each time you start it. the user.js i linked you to is an effort to make a user.js which protects the user's privacy and security. i dont use that one exactly but my user.js uses a lot of the preferences they have. 
another example of a user.js line can be:
`user_pref("layout.css.visited_links_enabled",false);`

**Comment by berrythesoftwarecodeprogrammar** [comment]
and yes it works, if you have noscript. it uses noscripts surrogate feature which lets you make preferences that run scripts on different pages. this script empties the plugins list on every domain except the ones you add to exceptions. you can test it here http://browserspy.dk/plugins.php
it works. you can see in my exceptions list i have .abc.net.au and .dailyliked.net because i let those sites see my plugin list (because they seem to rely on that to run flash). you could probably modify the preference or make a new one which can only expose flash to those sites instead of all plugins but i didnt bother to because i dont even encounter many sites which need plugins anymore.

also about the questions you asked gorhill, i think he might say that umatrix's job isnt to prevent fingerprinting, its only to prevent network requests. i use umatrix for that and use other addons/preferences to stop fingerprinting

**Comment by berrythesoftwarecodeprogrammar** [comment]
tl;dr: you can combat plugin enumeration by creating about:config preferences which use noscript's script surrogate feature. the same can also be achieved by adding said preferences to a user.js file which makes it easier to manage your own important preferences

**Comment by Roman-Nopantski** [comment]
@berrythesoftwarecodeprogrammar "prefs.js is where all about:config entries are stored".

Not exactly. It's where "user set" values from about:config are stored. These can be entries you edited in about:config and changed from default, or any prefs from user.js (the values can be default or not default) which are written to the prefs.js from user.js on a Firefox start. Anything in prefs.js is shown as "user set". It is also where extension preferences are stored. Prefs.js is definitely not all about:config entries - those are hardcoded or stored elsewhere; this is how a FF update can change pref default values, remove prefs, and add new prefs - all without messing with your prefs.js, which is in your profile. It's a profile file, not a application file.

tl;dr:
1. Preferences changed from default in about:config are written to prefs.js in real time
2. When Firefox starts, user.js preferences (default or not) are written to prefs.js
3. Then all prefs.js preferences are marked as "user_set" and their values overwritten in about:config for display and use
4. Topic for another day, locked prefs

**Comment by berrythesoftwarecodeprogrammar** [comment]
oh i see :) thanks for correcting me

**Comment by L-a-n-g-o-l-i-e-r-s** [comment]
@berrythesoftwarecodeprogrammar I am using the user script from the original user, I use NoScript in allow globally for the other protections and I don't know if that would work, I also think editing ignore list for scripts are probably easier. Thanks

**Comment by berrythesoftwarecodeprogrammar** [comment]
i use noscript that way too, surrogates work regardless of those settings :)

**Comment by L-a-n-g-o-l-i-e-r-s** [comment]
@berrythesoftwarecodeprogrammar is it as easy to add exceptions? good to know, thanks.

**Comment by berrythesoftwarecodeprogrammar** [comment]
1. ctrl+t (open a new tab)
2. type "about:config" enter
3. search "noplugin"
4. double click the `noscript.surrogate.noplugin.exceptions` entry
5. it brings up a popup with the exception list and you can just append the new site e.g. `.videosite.com`
6. press enter and close tab
